File: app/src/main/java/ch/threema/app/emojis/EmojiDrawable.java
package ch.threema.app.emojis;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class EmojiDrawable extends Drawable {
    private final SpriteCoordinates coordinates;
    private final int spritemapInSampleSize;
    private Bitmap bitmap;

    private static final Paint paint = new Paint(Paint.FILTER_BITMAP_FLAG | Paint.ANTI_ALIAS_FLAG);

    EmojiDrawable(SpriteCoordinates coordinates, int spritemapInSampleSize) {
        this.coordinates = coordinates;
        this.spritemapInSampleSize = spritemapInSampleSize;
    }

    @Override
    public int getIntrinsicWidth() {
        return spritemapInSampleSize * EmojiManager.EMOJI_WIDTH;
    }

    @Override
    public int getIntrinsicHeight() {
        return spritemapInSampleSize * EmojiManager.EMOJI_HEIGHT;
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        if (bitmap != null) {
            canvas.drawBitmap(bitmap,
                new Rect((int) ((float) (coordinates.x) / spritemapInSampleSize),
                    (int) ((float) coordinates.y / spritemapInSampleSize),
                    (int) ((float) (coordinates.x + EmojiManager.EMOJI_WIDTH) / spritemapInSampleSize),
                    (int) ((float) (coordinates.y + EmojiManager.EMOJI_HEIGHT) / spritemapInSampleSize)),
                getBounds(),
                paint);
        }
    }

    public void setBitmap(@Nullable Bitmap bitmap) {
        if (this.bitmap == null || !this.bitmap.sameAs(bitmap)) {
            this.bitmap = bitmap;
            invalidateSelf();
        }
    }

    @Override
    public void setAlpha(@IntRange(from = 0, to = 255) int alpha) {
    }

    @Override
    public void setColorFilter(@Nullable ColorFilter colorFilter) {
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSPARENT;
    }
}

File: app/src/main/java/ch/threema/app/emojis/EmojiDrawable.kt
package ch.threema.app.emojis

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.Paint
import android.graphics.PixelFormat
import android.graphics.Rect
import android.graphics.drawable.Drawable
import androidx.annotation.AnyThread
import androidx.annotation.IntRange
import ch.threema.app.utils.RuntimeUtil
import ch.threema.base.utils.getThreemaLogger
import kotlinx.coroutines.CompletableDeferred

private val logger = getThreemaLogger("EmojiDrawable")

class EmojiDrawable(
    private val coordinates: SpriteCoordinates,
    private val spritemapInSampleSize: Int,
) : Drawable() {
    private var bitmap: Bitmap? = null

    /**
     * Tracks whether the bitmap was already loaded. Will be set to 'completed' and then set to null to free up memory,
     * as once the bitmap is loaded it remains in this state.
     */
    private var loadedDeferred: CompletableDeferred<Unit>? = CompletableDeferred()

    override fun getIntrinsicWidth() =
        spritemapInSampleSize * EMOJI_WIDTH

    override fun getIntrinsicHeight() =
        spritemapInSampleSize * EMOJI_HEIGHT

    override fun draw(canvas: Canvas) {
        val bitmap = bitmap ?: return
        canvas.drawBitmap(
            bitmap,
            Rect(
                ((coordinates.x).toFloat() / spritemapInSampleSize).toInt(),
                (coordinates.y.toFloat() / spritemapInSampleSize).toInt(),
                ((coordinates.x + EMOJI_WIDTH).toFloat() / spritemapInSampleSize).toInt(),
                ((coordinates.y + EMOJI_HEIGHT).toFloat() / spritemapInSampleSize).toInt(),
            ),
            getBounds(),
            paint,
        )
    }

    @AnyThread
    fun setBitmap(bitmap: Bitmap) {
        if (this.bitmap != null) {
            logger.error("Bitmap already set for {}", coordinates)
            return
        }
        this.bitmap = bitmap
        loadedDeferred?.complete(Unit)
        loadedDeferred = null
        RuntimeUtil.runOnUiThread {
            invalidateSelf()
        }
    }

    /**
     * Indicates that the loading of the bitmap for this [EmojiDrawable] has failed, which will cause [awaitLoaded]
     * to throw an [EmojiLoadingException]. Will have no effect if the bitmap was already loaded.
     */
    @AnyThread
    fun setFailed() {
        loadedDeferred?.completeExceptionally(EmojiLoadingException())
    }

    /**
     * Suspends until the bitmap for this [EmojiDrawable] is loaded.
     *
     * @throws EmojiLoadingException if loading failed.
     */
    suspend fun awaitLoaded() {
        loadedDeferred?.await()
    }

    override fun setAlpha(@IntRange(from = 0, to = 255) alpha: Int) {
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
    }

    @Deprecated("Deprecated in Java")
    override fun getOpacity() =
        PixelFormat.TRANSPARENT

    class EmojiLoadingException : Exception()

    companion object {
        private const val EMOJI_HEIGHT = 64
        private const val EMOJI_WIDTH = 64

        private val paint = Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)
    }
}
